cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/metrics/MetricsHandlerTestRun.java

```diff
@@ -670,6 +738,221 @@ public void testSearchMetrics() throws Exception {
+    // test invalid request - query without any query Params and body content
+    HttpResponse response = doPost("/v3/metrics/query", null);
+    Assert.assertEquals(400, response.getStatusLine().getStatusCode());
+  }
+
+  JsonObject getSingleQueryJson(Map<String, String> tags, List<String> metrics,
+                               List<String> groupBy, Map<String, String> timeRange) {
+    JsonObject query = new JsonObject();
+    JsonObject tagsJson = new JsonObject();
+    for (Map.Entry<String, String> entry : tags.entrySet()) {
+      tagsJson.addProperty(entry.getKey(), entry.getValue());
+    }
+    query.add("tags", tagsJson);
+    JsonArray metricsArray = new JsonArray();
+    for (String metric : metrics) {
+      metricsArray.add(new JsonPrimitive(metric));
+    }
+    query.add("metrics", metricsArray);
+
+    JsonArray groupByArray = new JsonArray();
+    for (String group : groupBy) {
+      groupByArray.add(new JsonPrimitive(group));
+    }
+    query.add("groupBy", groupByArray);
+
+    JsonObject timerangeJson = new JsonObject();
+    for (Map.Entry<String, String> entry : timeRange.entrySet()) {
+      timerangeJson.addProperty(entry.getKey(), entry.getValue());
+    }
+    query.add("timeRange", timerangeJson);
+    return query;
+  }
+
+  class QueryResult {
+    Map<String, String> grouping;
+    String metricName;
+    long numPoints;
+    long totalSum;
+
+    public QueryResult(Map<String, String> grouping, String metricName, long numPoints, long totalSum) {
+      this.grouping = grouping;
+      this.metricName = metricName;
+      this.numPoints = numPoints;
+      this.totalSum = totalSum;
+    }
+
+    public long getTotalSum() {
+      return totalSum;
+    }
+
+    public long getNumPoints() {
+      return numPoints;
+    }
+
+    public String getMetricName() {
+      return metricName;
+    }
+
+    public Map<String, String> getGrouping() {
+      return grouping;
+    }
+  }
+
+
+  private  void batchTest(JsonObject batchRequests,
+                          ImmutableMap<String, ImmutableList<ImmutableList<QueryResult>>> expected) throws Exception {
+    String url = "/v3/metrics/query";
+    HttpURLConnection urlConn = openURL(MetricsSuiteTestBase.getEndPoint(url).toURL(),
+                                        HttpMethod.POST);
+    urlConn.setDoOutput(true);
+    String jsonBatch = new Gson().toJson(batchRequests);
+    urlConn.getOutputStream().write(jsonBatch.getBytes());
+    Assert.assertEquals(HttpResponseStatus.OK.getCode(), urlConn.getResponseCode());
+
+    String result = IOUtils.toString(urlConn.getInputStream(), Charsets.UTF_8);
+    Map<String, List<MetricQueryResult>> results =
+      new Gson().fromJson(result, new TypeToken<Map<String, List<MetricQueryResult>>>() {
+      }.getType());
+    urlConn.disconnect();
+    // check we have all the keys
+    Assert.assertEquals(expected.keySet(), results.keySet());
+    for (Map.Entry<String, List<MetricQueryResult>> entry : results.entrySet()) {
```

Comment:
we can have multiple queries under a single query-id , that's what i am testing here.